syntax = "proto3";

package stock;

service StockService {
  rpc AddProduct (AddProductRequest) returns (AddProductResponse);
  rpc ChangeQuantity (ChangeQuantityRequest) returns (ChangeQuantityResponse);
  rpc ListProducts (Empty) returns (ProductList);
  rpc Exit (Empty) returns (ExitResponse);
}

message AddProductRequest {
  string description = 1;
  int32 quantity = 2;
}

message AddProductResponse {
  int32 product_id = 1;
}

message ChangeQuantityRequest {
  int32 product_id = 1;
  int32 value = 2;
}

message ChangeQuantityResponse {
  int32 status = 1; // -1 for insufficient stock, -2 for invalid product ID
}

message Product {
  int32 product_id = 1;
  string description = 2;
  int32 quantity = 3;
}

message ProductList {
  repeated Product products = 1;
}

message ExitResponse {
  int32 total_products = 1;
}

message Empty {}
